runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryStrategy.kt

```diff
@@ -0,0 +1,160 @@
+        return result.getOrNull()!!
+    }
+
+    /**
+     * Handles the termination of the retry loop because of a non-retryable failure by marking the [RetryToken] as
+     * failed and throwing a [RetryFailureException].
+     * @param token The [RetryToken] used in the attempt that was unsuccessful.
+     * @param attempt The number of attempts completed.
+     * @param result The [Result] that yielded the non-retryable condition.
+     */
+    private suspend fun <R> throwFailure(token: RetryToken, attempt: Int, result: Result<R>): Nothing {
+        token.notifyFailure()
+        throw RetryFailureException(
+            "The operation failed",
+            result.exceptionOrNull(),
+            attempt,
+            result.getOrNull(),
+        )
+    }
+
+    /**
+     * Handles the termination of the retry loop because too much time has elapsed by marking the [RetryToken] as failed
+     * and throwing a [TimedOutException].
+     * @param token The [RetryToken] used in the attempt that was waiting or executing when the timeout occurred.
+     * @param attempt The number of attempts completed.
+     * @param previousResult The last result that was received (i.e., from the prior loop iteration).
+     */
+    private suspend fun <R> throwTimeOut(token: RetryToken, attempt: Int, previousResult: Result<R>?): Nothing {
+        token.notifyFailure()
+        throw TimedOutException(
+            "Took more than ${options.maxTimeMs}ms to yield a result",
+            attempt,
+            previousResult?.getOrNull(),
+            previousResult?.exceptionOrNull(),
+        )
+    }
+
+    /**
+     * Handles the termination of the retry loop because too many attempts have been made by marking the [RetryToken] as
+     * failed and throwing a [TimedOutException].
+     * @param token The [RetryToken] used in the attempt that was unsuccessful.
+     * @param attempt The number of attempts completed.
+     * @param result The [Result] that yielded a retryable condition (but which won't be retried because we've already
+     * tried too many times).
+     */
+    private suspend fun <R> throwTooManyAttempts(token: RetryToken, attempt: Int, result: Result<R>): Nothing {
+        token.notifyFailure()
+        throw TooManyAttemptsException(
+            "Took more than ${options.maxAttempts} to get a successful response",
+            attempt,
+            result.getOrNull(),
+            result.exceptionOrNull(),
+        )
+    }
+}
+
+/**
+ * Defines configuration for a [StandardRetryStrategy].
+ * @param maxTimeMs The maximum amount of time to retry (in milliseconds).
+ * @param maxAttempts The maximum number of attempts to make (including the first attempt).
+ */
+data class StandardRetryStrategyOptions(val maxTimeMs: Long, val maxAttempts: Int) {
+    companion object {
+        /**
+         * The default retry strategy configuration.
+         */
+        val Default = StandardRetryStrategyOptions(maxTimeMs = 10_000, maxAttempts = 10)
```

Comment:
question

Where do these values come from? The SEP indicates 3 as default for max attempts.

Reply:
They are my own attempt at establishing sensible defaults. I can update to the SEP-specified values.